<h2>Compilation Error</h2><pre>
main.cu(13): error: identifier "uint32_t" is undefined
      uint32_t len;
      ^

main.cu(14): error: identifier "uint32_t" is undefined
      uint32_t offset;
      ^

main.cu(17): error: incomplete type "void" is not allowed
  __attribute__((device)) void radix_sort_shared(uint32_t* data, uint32_t* temp, uint32_t len) {
                               ^

main.cu(17): error: identifier "uint32_t" is undefined
  __attribute__((device)) void radix_sort_shared(uint32_t* data, uint32_t* temp, uint32_t len) {
                                                 ^

main.cu(17): error: identifier "temp" is undefined
  __attribute__((device)) void radix_sort_shared(uint32_t* data, uint32_t* temp, uint32_t len) {
                                                                           ^

main.cu(17): error: expected a ")"
  __attribute__((device)) void radix_sort_shared(uint32_t* data, uint32_t* temp, uint32_t len) {
                                                                                          ^

main.cu(17): error: expected a ";"
  __attribute__((device)) void radix_sort_shared(uint32_t* data, uint32_t* temp, uint32_t len) {
                                                                                               ^

main.cu(38): warning #12-D: parsing restarts here after previous syntax error
              data[i] = temp[i];
                               ^

Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"

main.cu(39): error: expected a declaration
          }
          ^

main.cu(64): warning #12-D: parsing restarts here after previous syntax error
              data[i] = temp[i];
                               ^

main.cu(65): error: expected a declaration
          }
          ^

main.cu(103): warning #12-D: parsing restarts here after previous syntax error
      radix_sort_shared(local_data, local_temp, len);
                                                    ^

main.cu(105): error: explicit type is missing ("int" assumed)
      __syncthreads();
      ^

main.cu(105): error: cannot overload functions distinguished by return type alone
      __syncthreads();
      ^

main.cu(107): error: expected a declaration
      for (uint32_t i = 0; i < len; ++i) {
      ^

main.cu(112): warning #12-D: parsing restarts here after previous syntax error
  vector<uint32_t> data_store;
                             ^

main.cu(123): error: expected a ";"
      uint32_t offset = 0;
               ^

main.cu(131): error: expected a ";"
              uint32_t n = stoul(chunk);
                       ^

main.cu(132): error: "uint32_t" is not a type name
              vector<uint32_t> values;
                     ^

/usr/include/c++/13/bits/ptr_traits.h(168): error: class "std::__replace_first_arg<<error-type> *, const <error-type>>" has no member "type"
        __rebind, _Ptr, _Up>::type;
                              ^
          detected during:
            instantiation of type "std::__ptr_traits_impl<_Ptr, _Elt>::rebind<const <error-type>> [with _Ptr=<error-type> *, _Elt=std::__undefined]" at line 124 of /usr/include/c++/13/bits/alloc_traits.h
            instantiation of class "std::allocator_traits<_Alloc>::_Ptr<_Func, _Tp, <unnamed>> [with _Alloc=std::allocator<<error-type>>, _Func=std::__allocator_traits_base::__c_pointer, _Tp=const <error-type>, <unnamed>=void]" at line 157 of /usr/include/c++/13/bits/alloc_traits.h
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::allocator<<error-type>>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>]" at line 88 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/ptr_traits.h(168): error: class "std::__replace_first_arg<<error-type> *, void>" has no member "type"
        __rebind, _Ptr, _Up>::type;
                              ^
          detected during:
            instantiation of type "std::__ptr_traits_impl<_Ptr, _Elt>::rebind<void> [with _Ptr=<error-type> *, _Elt=std::__undefined]" at line 124 of /usr/include/c++/13/bits/alloc_traits.h
            instantiation of class "std::allocator_traits<_Alloc>::_Ptr<_Func, _Tp, <unnamed>> [with _Alloc=std::allocator<<error-type>>, _Func=std::__allocator_traits_base::__v_pointer, _Tp=void, <unnamed>=void]" at line 165 of /usr/include/c++/13/bits/alloc_traits.h
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::allocator<<error-type>>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>]" at line 88 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/ptr_traits.h(168): error: class "std::__replace_first_arg<<error-type> *, const void>" has no member "type"
        __rebind, _Ptr, _Up>::type;
                              ^
          detected during:
            instantiation of type "std::__ptr_traits_impl<_Ptr, _Elt>::rebind<const void> [with _Ptr=<error-type> *, _Elt=std::__undefined]" at line 124 of /usr/include/c++/13/bits/alloc_traits.h
            instantiation of class "std::allocator_traits<_Alloc>::_Ptr<_Func, _Tp, <unnamed>> [with _Alloc=std::allocator<<error-type>>, _Func=std::__allocator_traits_base::__cv_pointer, _Tp=const void, <unnamed>=void]" at line 173 of /usr/include/c++/13/bits/alloc_traits.h
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::allocator<<error-type>>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>]" at line 88 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/alloc_traits.h(57): error: class "std::__replace_first_arg<std::allocator<<error-type>>, <error-type>>" has no member "type"
     typename __replace_first_arg<_Tp, typename _Tp::value_type>::type,
                                                                  ^
          detected during:
            instantiation of class "std::__allocator_traits_base::__rebind<_Tp, _Up, <unnamed>> [with _Tp=std::allocator<<error-type>>, _Up=<error-type>, <unnamed>=void]" at line 95
            instantiation of type "std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>" at line 228
            instantiation of type "std::allocator_traits<_Alloc>::rebind_alloc<<error-type>> [with _Alloc=std::allocator<<error-type>>]" at line 126 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>>::rebind<_Tp> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>, _Tp=<error-type>]" at line 88 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/alloc_traits.h(95): error: class "std::__allocator_traits_base::__rebind<std::allocator<<error-type>>, <error-type>, void>" has no member "type"
        = typename __allocator_traits_base::template __rebind<_Alloc, _Up>::type;
                                                                            ^
          detected during:
            instantiation of type "std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>" at line 228
            instantiation of type "std::allocator_traits<_Alloc>::rebind_alloc<<error-type>> [with _Alloc=std::allocator<<error-type>>]" at line 126 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>>::rebind<_Tp> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>, _Tp=<error-type>]" at line 88 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/ext/alloc_traits.h(44): error: a class or namespace qualified name is required
  template<typename _Alloc, typename = typename _Alloc::value_type>
                                                ^
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <<error-type>> at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/ext/alloc_traits.h(44): error: expected a ","
  template<typename _Alloc, typename = typename _Alloc::value_type>
                                                      ^
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <<error-type>> at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/alloc_traits.h(110): error: a class or namespace qualified name is required
        typedef typename _Alloc::value_type value_type;
                         ^
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=<error-type>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=<error-type>, <unnamed>=<error-type>]" at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/alloc_traits.h(110): error: global-scope qualifier (leading "::") is not allowed
        typedef typename _Alloc::value_type value_type;
                               ^
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=<error-type>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=<error-type>, <unnamed>=<error-type>]" at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/alloc_traits.h(110): error: expected a ";"
        typedef typename _Alloc::value_type value_type;
                                            ^
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=<error-type>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=<error-type>, <unnamed>=<error-type>]" at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/alloc_traits.h(117): error: identifier "value_type" is undefined
        using pointer = __detected_or_t<value_type*, __pointer, _Alloc>;
                                        ^
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=<error-type>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=<error-type>, <unnamed>=<error-type>]" at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/alloc_traits.h(157): error: expected a ">"
        using const_pointer = typename _Ptr<__c_pointer, const value_type>::type;
                                                               ^
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=<error-type>]" at line 47 of /usr/include/c++/13/ext/alloc_traits.h
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=<error-type>, <unnamed>=<error-type>]" at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/ext/alloc_traits.h(53): error: class "std::allocator_traits<<error-type>>" has no member "value_type"
      typedef typename _Base_type::value_type value_type;
                                   ^
          detected during:
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=<error-type>, <unnamed>=<error-type>]" at line 89 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/bits/stl_vector.h(134): error: not a class or struct name
   : public _Tp_alloc_type, public _Vector_impl_data
            ^
          detected during:
            instantiation of class "std::_Vector_base<_Tp, _Alloc>::_Vector_impl [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 374
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 428
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/ext/alloc_traits.h(44): error: a class or namespace qualified name is required
  template<typename _Alloc, typename = typename _Alloc::value_type>
                                                ^
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <<error-type>> at line 450 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

/usr/include/c++/13/ext/alloc_traits.h(44): error: expected a ","
  template<typename _Alloc, typename = typename _Alloc::value_type>
                                                      ^
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <<error-type>> at line 450 of /usr/include/c++/13/bits/stl_vector.h
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" at line 132 of main.cu

main.cu(133): error: identifier "n" is undefined
              values.reserve(n);
                             ^

main.cu(135): error: expected a ";"
              for (uint32_t i = 0; i < n && getline(ss, chunk, ','); ++i)
                            ^

main.cu(135): error: i is not a template
              for (uint32_t i = 0; i < n && getline(ss, chunk, ','); ++i)
                                   ^

main.cu(135): error: expected a ">"
              for (uint32_t i = 0; i < n && getline(ss, chunk, ','); ++i)
                                                                   ^

main.cu(135): error: identifier "i" is undefined
              for (uint32_t i = 0; i < n && getline(ss, chunk, ','); ++i)
                                   ^

main.cu(139): error: identifier "offset" is undefined
                  all_seqs.push_back({n, offset});
                                         ^

main.cu(140): error: identifier "data_store" is undefined
                  data_store.insert(data_store.end(), values.begin(), values.end());
                  ^

/usr/include/c++/13/bits/stl_iterator.h(1062): error: class "std::iterator_traits<<error-type> *>" has no member "iterator_category"
        typedef typename __traits_type::iterator_category iterator_category;
                                        ^
          detected during instantiation of class "__gnu_cxx::__normal_iterator<_Iterator, _Container> [with _Iterator=<error-type> *, _Container=std::vector<<error-type>, std::allocator<<error-type>>>]" at line 140 of main.cu

/usr/include/c++/13/bits/stl_iterator.h(1063): error: class "std::iterator_traits<<error-type> *>" has no member "value_type"
        typedef typename __traits_type::value_type value_type;
                                        ^
          detected during instantiation of class "__gnu_cxx::__normal_iterator<_Iterator, _Container> [with _Iterator=<error-type> *, _Container=std::vector<<error-type>, std::allocator<<error-type>>>]" at line 140 of main.cu

/usr/include/c++/13/bits/stl_iterator.h(1064): error: class "std::iterator_traits<<error-type> *>" has no member "difference_type"
        typedef typename __traits_type::difference_type difference_type;
                                        ^
          detected during instantiation of class "__gnu_cxx::__normal_iterator<_Iterator, _Container> [with _Iterator=<error-type> *, _Container=std::vector<<error-type>, std::allocator<<error-type>>>]" at line 140 of main.cu

/usr/include/c++/13/bits/stl_iterator.h(1065): error: class "std::iterator_traits<<error-type> *>" has no member "reference"
        typedef typename __traits_type::reference reference;
                                        ^
          detected during instantiation of class "__gnu_cxx::__normal_iterator<_Iterator, _Container> [with _Iterator=<error-type> *, _Container=std::vector<<error-type>, std::allocator<<error-type>>>]" at line 140 of main.cu

/usr/include/c++/13/bits/stl_iterator.h(1066): error: class "std::iterator_traits<<error-type> *>" has no member "pointer"
        typedef typename __traits_type::pointer pointer;
                                        ^
          detected during instantiation of class "__gnu_cxx::__normal_iterator<_Iterator, _Container> [with _Iterator=<error-type> *, _Container=std::vector<<error-type>, std::allocator<<error-type>>>]" at line 140 of main.cu

main.cu(156): error: expected a ";"
      uint32_t i = 0;
               ^

main.cu(159): error: expected a ";"
          for (uint32_t j = 0; j < s.len; ++j)
                        ^

main.cu(159): error: j is not a template
          for (uint32_t j = 0; j < s.len; ++j)
                               ^

main.cu(159): error: expected a ">"
          for (uint32_t j = 0; j < s.len; ++j)
                                        ^

main.cu(159): error: identifier "j" is undefined
          for (uint32_t j = 0; j < s.len; ++j)
                               ^

main.cu(160): error: identifier "data_store" is undefined
              f << "," << data_store[i++];
                          ^

main.cu(160): error: identifier "i" is undefined
              f << "," << data_store[i++];
                                     ^

main.cu(186): error: identifier "dev_data" is undefined
      uint32_t* dev_data;
                ^

main.cu(187): error: identifier "temp_data" is undefined
      uint32_t* temp_data;
                ^

main.cu(190): error: identifier "data_store" is undefined
      size_t total_bytes = data_store.size() * sizeof(uint32_t);
                           ^

main.cu(206): error: expected a ";"
      uint32_t max_seq_len = 0;
               ^

main.cu(208): error: identifier "max_seq_len" is undefined
          max_seq_len = max(max_seq_len, seq.len);
          ^

main.cu(216): error: expected a ";"
              uint32_t copy_from = all_seqs[done].offset;
                       ^

main.cu(217): error: expected a ";"
              uint32_t copy_bytes = 0;
                       ^

main.cu(219): error: expected a ";"
              uint32_t batch_max_len = 0;
                       ^

main.cu(221): error: identifier "batch_max_len" is undefined
                  batch_max_len = max(batch_max_len, all_seqs[done + j].len);
                  ^

main.cu(225): error: expected a ";"
                  uint32_t limit = all_seqs[done].offset;
                           ^

main.cu(227): error: identifier "limit" is undefined
                      limit = all_seqs[done + j].offset + all_seqs[done + j].len;
                      ^

main.cu(228): error: identifier "copy_bytes" is undefined
                  copy_bytes = (limit - copy_from) * sizeof(uint32_t);
                  ^

main.cu(228): error: identifier "limit" is undefined
                  copy_bytes = (limit - copy_from) * sizeof(uint32_t);
                                ^

main.cu(228): error: identifier "copy_from" is undefined
                  copy_bytes = (limit - copy_from) * sizeof(uint32_t);
                                        ^

main.cu(241): error: identifier "batch_max_len" is undefined
              if (batch_max_len <= SHARED_MEM_THRESHOLD) {
                  ^

main.cu(243): error: identifier "radix_sort_shared_kernel" is undefined
                  radix_sort_shared_kernel<<<grid, block, shared_mem_size, streams[s]>>>(
                  ^

main.cu(246): error: identifier "radix_sort_batch" is undefined
                  radix_sort_batch<<<grid, block, 0, streams[s]>>>(
                  ^

main.cu(250): error: identifier "copy_from" is undefined
              cudaMemcpyAsync(data_store.data() + copy_from,
                                                  ^

main.cu(252): error: identifier "copy_bytes" is undefined
                              copy_bytes,
                              ^

69 errors detected in the compilation of "main.cu".

</pre>